File: Practica6/Python/Vector.py
import math
class Vector:
    def __init__(self, x, y, z):
        self.x, self.y, self.z = x, y, z
    def anadir(self, o):
        return Vector(self.x + o.x, self.y + o.y, self.z + o.z)
    def restaVectorial(self, o):
        return Vector(self.x - o.x, self.y - o.y, self.z - o.z)
    def multiplicacionEscalar(self, scalar):
        return Vector(self.x * scalar, self.y * scalar, self.z * scalar)
    def magnitude(self):
        return math.sqrt(self.x ** 2 + self.y ** 2 + self.z ** 2)
    def escalar(self, o):
        return self.x * o.x + self.y * o.y + self.z * o.z
    def cruz(self, o):
        return Vector(
            self.y * o.z - self.z * o.y,
            self.z * o.x - self.x * o.z,
            self.x * o.y - self.y * o.x
        )
    def perpendicular(self, o):
        return self.escalar(o) == 0 
    def paralelo(self, o):
        return self.cruz(o).magnitude() == 0
    def proyeccion(self, o):
        factor = self.escalar(o) / o.escalar(o)
        return o.multiplicacionEscalar(factor)
    def componentes(self, o):
        return self.escalar(o) / o.magnitude()
a = Vector(2, 5, 0)
b = Vector(4, 10, 0)
proyeccion = a.proyeccion(b)
print("Proyección de A sobre B:", "x:", proyeccion.x, " y:", proyeccion.y, " z:", proyeccion.z)
print("Componente de A en B:", a.componentes(b))
print("El escalar es: ", a.escalar(b))
print("El producto vectorial: ", a.cruz(b).x, a.cruz(b).y, a.cruz(b).z)
print("¿Son perpendiculares?", a.perpendicular(b))
print("¿Son paralelos?", a.paralelo(b))

File: Practica6/Java/Vector.java
class Vector {
    double x;
    double y;
    double z;
    Vector(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    Vector añadir(Vector o) {
        return new Vector(x + o.x, y + o.y, z + o.z);
    }
    Vector restaVector(Vector o) {
        return new Vector(x - o.x, y - o.y, z - o.z);
    }
    Vector multiplicaVector(double s) {
        return new Vector(x * s, y * s, z * s);
    }
    double magnitude() {
        return Math.sqrt(x * x + y * y + z * z);
    }
    double productoEscalar(Vector o) {
        return x * o.x + y * o.y + z * o.z;
    }
    Vector cruz(Vector o) {
        return new Vector(
            y * o.z - z * o.y,
            z * o.x - x * o.z,
            x * o.y - y * o.x
        );
    }
    boolean perpendicular(Vector o) {
        return productoEscalar(o) == 0 || añadir(o).magnitude() == restaVector(o).magnitude();
    }
    boolean paralelo(Vector o) {
        return cruz(o).magnitude() == 0;
    }
    Vector proyeccion(Vector o) {
        double f = productoEscalar(o) / o.productoEscalar(o);
        return o.multiplicaVector(f);
    }
    double componentes(Vector o) {
        return productoEscalar(o) / o.magnitude();
    }
}